src/Storages/Distributed/DistributedSink.cpp

```diff
@@ -372,10 +370,11 @@ DistributedSink::runWritingJob(JobReplica & job, const Block & current_block, si
                         throw Exception(ErrorCodes::LOGICAL_ERROR, "There are several writing job for an automatically replicated shard");
 
                     /// TODO: it make sense to rewrite skip_unavailable_shards and max_parallel_replicas here
-                    auto results = shard_info.pool->getManyChecked(timeouts, settings, PoolMode::GET_ONE, main_table.getQualifiedName());
-                    if (results.empty() || results.front().entry.isNull())
-                        throw Exception(ErrorCodes::LOGICAL_ERROR, "Expected exactly one connection for shard {}", toString(job.shard_index));
-
+                    /// NOTE: INSERT will also take into account max_replica_delay_for_distributed_queries
+                    /// (anyway fallback_to_stale_replicas_for_distributed_queries=true by default)
+                    auto results = shard_info.pool->getManyChecked(timeouts, settings, PoolMode::GET_ONE, storage.remote_storage.getQualifiedName());
```

Comment:
Unfortunately, it won't work. Here we'll get only one connection in `results` since `PoolMode::GET_ONE`, and it can be to read-only replica. So, it looks like having a setting `skip_read_only_replicas` is the simplest way to implement avoidance of the read-only replicas during distributed insert.
Yes, it's different from `prefer_non_read_only_replicas`, but it's a bit simpler from an implementation standpoint and has more deterministic outcome (not completely though, since replica can become read-only anytime).

So, I'd suggest adding just one new setting `skip_read_only_replicas` instead of introducing enum setting.

P.S. I apologize for confusion if any, - it seems we're going back and forth, but at least now I have a more clear POV.

Reply:
>Unfortunately, it won't work. Here we'll get only one connection in results since PoolMode::GET_ONE, and it can be to read-only replica

That is correct, in this version of the patch it does not work, while it works in one of the intermediate versions, when the sorting of the shards was done inside the pool.

>So, I'd suggest adding just one new setting skip_read_only_replicas instead of introducing enum setting.

OK, let's go this way.

Reply:
Done